assets/app/scripts/services/applicationGenerator.js

```diff
@@ -0,0 +1,267 @@
+"use strict";
+
+angular.module("openshiftConsole")
+
+  .service("ApplicationGenerator", function(){
+    var osApiVersion = "v1beta1";
+    var k8eApiVersion = "v1beta3";
+    
+    var service = {};
+    var scope = service;
+    
+    scope._generateSecret_ = function(){
+        //http://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
+        function s4() {
+          return Math.floor((1 + Math.random()) * 0x10000)
+              .toString(16)
+              .substring(1);
+          }
+        return s4()+s4()+s4()+s4();
+      };
+
+    /**
+    * Find the 'first' port of exposed ports.
+    * @param            ports  list of ports (e.g {containerPort: 80, protocol: "tcp"})
+    * @return {integer} The port/protocol pair of the lowest conainer port
+    */
+    scope._getFirstPort_ = function(ports){
+      var first = "None";
+      ports.forEach(function(port){
+        if(first === "None"){
+            first = port;
+          }else{
+            if(port.containerPort < first.containerPort){
+              first = port;
+            }
+          }
+        }
+      );
+      return first;
+    };
+    
+    /**
+     * Generate resource definitions to support the given input
+     * @param {type} input
+     * @returns Hash of resource definitions
+     */
+    scope.generate = function(input){
+      //map ports to k8e structure
+      var ports = [];
+      angular.forEach(input.image.dockerImageMetadata.ContainerConfig.ExposedPorts, function(value, key){
+        var parts = key.split("/");
+        ports.push(
+          {
+            containerPort: parseInt(parts[0]), 
+            name: input.name + "-" + parts[1] + "-" + parts[0],
+            protocol: parts[1]
+          });
+      });
+
+      //augment labels
+      input.labels.name = input.name;
+      input.labels.generatedby = "OpenShiftWebConsole";
+      
+      var imageSpec;
+      if(input.buildConfig.sourceUrl !== null){
+        imageSpec = {
+          name: input.name, 
+          tag: "latest",
+          toString: function(){
+            return this.name + ":" + this.tag;
+          }
+        };
+      }
+      
+      var resources = {
+        imageRepo: scope._generateImageRepo_(input),
+        buildConfig: scope._generateBuildConfig_(input, imageSpec, input.labels),
+        deploymentConfig: scope._generateDeploymentConfig_(input, imageSpec, ports, input.labels),
+        service: scope._generateService_(input, input.name, scope._getFirstPort_(ports))
+      };
+      resources.route = scope._generateRoute_(input, resources.service.metadata.name, resources.service.metadata.name);
```

Comment:
@csrwng One issue here is its named wrong, should be input.name as one arg...but it is called out separately here because I am using the service name as input to the function.